mono/metadata/assembly.c

```diff
@@ -1251,15 +1251,21 @@ mono_stringify_assembly_name (MonoAssemblyName *aname)
 {
 	const char *quote = (aname->name && g_ascii_isspace (aname->name [0])) ? "\"" : "";
 
-	return g_strdup_printf (
-		"%s%s%s, Version=%d.%d.%d.%d, Culture=%s, PublicKeyToken=%s%s",
-		quote, aname->name, quote,
-		aname->major, aname->minor, aname->build, aname->revision,
-		aname->culture && *aname->culture? aname->culture: "neutral",
-		aname->public_key_token [0] ? (char *)aname->public_key_token : "null",
-		(aname->flags & ASSEMBLYREF_RETARGETABLE_FLAG) ? ", Retargetable=Yes" : "");
+	char *version_string = "";
+	char *culture_string = "";
+	char *token_string = "";
+
+	if (aname->culture [0])
+		culture_string = g_strdup_printf (", Culture=%s", aname->culture);
+	if (aname->public_key_token [0])
+		token_string = g_strdup_printf (", PublicKeyToken=%s%s", aname->public_key_token,(aname->flags & ASSEMBLYREF_RETARGETABLE_FLAG) ? ", Retargetable=Yes" : "");
+	if (aname->has_version)
+		version_string = g_strdup_printf (", Version=%d.%d.%d.%d", aname->major, aname->minor, aname->build, aname->revision);
+
+	return g_strdup_printf ("%s%s%s%s%s%s", quote, aname->name, quote, version_string, culture_string, token_string);
```

Comment:
There's a problem here: each call to `g_strdup_printf` allocates memory that needs to be freed.  So one option would be to make sure to call `g_free` on each of `version_string`, `culture_string` and `token_string` before returning.

But a cleaner way is to use the `GString` API:
```c
   GString *str;
   str = g_string_new ();
   g_string_append_printf (str, "%s%s%s", quote, aname->name, quote);
   if (aname->culture[0])
        g_string_append_printf (str, ", Culture=%s", aname->culture);
   ...
   if (aname->has_version)
        g_string_append_printf (", Version=%d.%d.%d.%d", aname->major, ..., aname->revision);

   char *result = g_string_free (str, FALSE); /* result is the final formatted string */
   return result;
```

Reply:
Thank you ! But little mistake in g_string_new () -  
need to pass to the constructor   gchar * . In this case, I think the best is NULL